src/disco/bank/fd_txncache.c

```diff
@@ -0,0 +1,741 @@
+#include "fd_txncache.h"
+
+#define SORT_NAME        sort_slot_ascend
+#define SORT_KEY_T       ulong
+#define SORT_BEFORE(a,b) (a)<(b)
+#include "../../util/tmpl/fd_sort.c"
+
+/* The number of transactions in each page.  This needs to be high
+   enough to amoritze the cost of caller code reserving pages from,
+   and returning pages to the pool, but not so high that the memory
+   wasted from blockhashes with only one transaction is significant. */
+
+#define FD_TXNCACHE_TXNS_PER_PAGE (16384UL)
+
+/* The number of unique entries in the hash lookup table for each
+   blockhash.  A higher value here uses more memory but enables faster
+   lookups. */
+
+#define FD_TXNCACHE_BLOCKCACHE_MAP_CNT (524288UL)
+
+/* The number of unique entries in the hash lookup table for each
+   (slot, blockhash).  This prevents all the entries needing to be in
+   one slotcache list, so insertions can happen concurrently. */
+
+#define FD_TXNCACHE_SLOTCACHE_MAP_CNT (1024UL)
+
+struct fd_txncache_private_txn {
+  uint  blockcache_next; /* Pointer to the next element in the blockcache hash chain containing this entry from the pool. */
+  uint  slotblockcache_next;  /* Pointer to the next element in the slotcache hash chain containing this entry from the pool. */
+
+  ulong slot;            /* Slot that the transaction was executed.  A transaction might be in the cache
+                            multiple times if it was executed in a different slot on different forks.  The
+                            same slot will not appear multiple times however. */
+  uchar txnhash[ 20 ];   /* The transaction hash, truncated to 20 bytes.  The hash is not always the first 20
+                            bytes, but is 20 bytes starting at some arbitrary offset given by the key_offset value
+                            of the containing by_blockhash entry. */
+  uchar result[ 40 ];    /* The result of executing the transaction. */
+};
+
+typedef struct fd_txncache_private_txn fd_txncache_private_txn_t;
+
+struct fd_txncache_private_txnpage {
```

Comment:
You should mention explicitly here that all the txns refer to the same blockhash